### PR TITLE
Fixed Driver Dependency Tests on Windows

### PR DESCRIPTION
**Pull request by gmittert**
All Driver/Dependency tests now pass on Windows!

Requires https://github.com/apple/swift/pull/22596

```
-- Testing: 52 of 4486 tests, 12 threads --                                                                                   
Testing: 0 .. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90..                                                                    
Testing Time: 150.70s                                                                                                         
  Expected Passes    : 43                                                                                                     
  Expected Failures  : 2                                                                                                      
  Unsupported Tests  : 7 
```
The two expected failures are due to tests which expect parallel execution. The unsupported tests are due to the use of `$(subshells)`, mostly to get the version of swift compiler that is running.

As part of this, I've also implemented output buffering for the default TaskQueue implementation since it was required for a few of the tests. Are there platforms besides Windows that use the default implementation that I need to be aware of?  @compnerd had mentioned rewriting the Windows TasksQueue implementation so that it supports parallel execution soon, so this might be a temporary thing anyways.

@jrose-apple

### CODE REVIEW

File: lib/Basic/Default/TaskQueue.inc
Comment (jrose-apple) [inline lib/Basic/Default/TaskQueue.inc]: Comment needs updating!

File: test/Driver/Dependencies/embed-bitcode-parallel.swift
Comment (jrose-apple) [inline test/Driver/Dependencies/embed-bitcode-parallel.swift]: Why the change in behavior here?

File: test/Driver/Dependencies/one-way-parallel.swift
Comment (jrose-apple) [inline test/Driver/Dependencies/one-way-parallel.swift]: This backslash isn't going to bind correctly, is it?

File: lib/Basic/Default/TaskQueue.inc
Comment (jrose-apple) [inline lib/Basic/Default/TaskQueue.inc]: No need to do any of this type conversion, or to use a SmallVector instead of a plain array. There's an implicit conversion from SmallString (a subclass of SmallVector) to StringRef, and from StringRef to Optional\<StringRef>. So you should be able to just pass `stdoutPath` and `stderrPath` directly here. (We've also brought None into scope, so you don't need the `llvm::` qualification.)

File: lib/Basic/Default/TaskQueue.inc
Comment (jrose-apple) [inline lib/Basic/Default/TaskQueue.inc]: This is clever, but what deletes the files after the run is finished?

File: lib/Basic/Default/TaskQueue.inc
Comment (jrose-apple) [inline lib/Basic/Default/TaskQueue.inc]: This is just `stdoutBuffer.get()->getBuffer()`. :-)

**Review by jrose-apple — COMMENTED** [review]
Clever! But I'm worried about the files being left behind on the filesystem—we'll have to do something about that.

File: lib/Basic/Default/TaskQueue.inc
Comment (jmittert) [reply lib/Basic/Default/TaskQueue.inc]: Hrm, I think ideally it would register with the compilation via `addTemporaryFile` but that would involve passing the compilation context all the way down which seem unideal. Would just removing the temp files after calling the callback be reasonable? Perhaps also registering with `RemoveFileOnSignal()`?

File: lib/Basic/Default/TaskQueue.inc
Comment (jmittert) [reply lib/Basic/Default/TaskQueue.inc]: That way is much more pleasant!

File: lib/Basic/Default/TaskQueue.inc
Comment (jrose-apple) [reply lib/Basic/Default/TaskQueue.inc]: :-/ Maybe? I suppose we don't expect them to be so big (on the offchance we can't clean them up), and they *are* going to be in a system temporary directory.

**Comment by jmittert** [comment]
Cleaned up both the code and the temp files!

File: lib/Basic/Default/TaskQueue.inc
Comment (jrose-apple) [inline lib/Basic/Default/TaskQueue.inc]: Ah, this is where to use SmallString, so that you don't need to explicitly make StringRefs later!

**Comment by jmittert** [comment]
TIL, list initialization is really neat.
```
     SmallVector<Optional<StringRef>, 4> redirects = {None, {stdoutPath}, {stderrPath}};
```
They did need the extra braces though, clang wasn't able to go from the stdoutPath: SmallString directly to Optional<StringRef>

File: lib/Basic/Default/TaskQueue.inc
Comment (jrose-apple) [inline lib/Basic/Default/TaskQueue.inc]: Last nitpick: This repo has an 80-column limit. But you can solve that anyway by using a plain fixed-sized array instead of a SmallVector.

```cxx
Optional<StringRef> redirects[] = {None, {stdoutPath}, {stderrPath}};
```

Comment (jmittert) [reply lib/Basic/Default/TaskQueue.inc]: Fixed.

**Comment by jrose-apple** [comment]
(not running full tests because the code being changed isn't used on POSIXy platforms…)

@swift-ci Please smoke test

**Review by jrose-apple — APPROVED** [review]

**Comment by jrose-apple** [comment]
Oops, I forgot that this one requires the other one.

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
Looks like something's not quite right even after merging the other PR. Can you take a look?

**Comment by jmittert** [comment]
I'm not entirely sure what happened. I'm unable to reproduce it on either a Linux (CentOS) or macOS machine locally. Is there something special about the Jenkins environment that I should know about?

The expected stdout and stderr look okay: the ls error is expected, but fake-ld.py never ran. The test overrides ld with fake-ld.py then expects swift to call it which then does a few assertions and prints out "Handled link" It seems like one of the tasks the driver spawned got killed before fake-ld was able to print...

```
13:39:28 --
13:39:28 Exit Code: 1
13:39:28 
13:39:28 Command Output (stdout):
13:39:28 --
13:39:28 /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/Driver/Output/filelists.swift.tmp/tmp/outputs-bee66b
13:39:28 /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/Driver/Output/filelists.swift.tmp/tmp/sources-7c25b1
13:39:28 /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/Driver/Output/filelists.swift.tmp/tmp-crash/outputs-ce9afc
13:39:28 /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/Driver/Output/filelists.swift.tmp/tmp-crash/sources-f43471
13:39:28 
13:39:28 --
13:39:28 Command Output (stderr):
13:39:28 --
13:39:28 ls: /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/Driver/Output/filelists.swift.tmp/tmp-fail/sources-*: No such file or directory
13:39:28 ls: /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/Driver/Output/filelists.swift.tmp/tmp-fail/outputs-*: No such file or directory
13:39:28 

<unknown>:0: error: compile command failed due to signal 9 (use -v to see invocation)
13:39:28 
/Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/swift/test/Driver/filelists.swift:68:16: error: CHECK-LINK: expected string not found in input
13:39:28 // CHECK-LINK: Handled link
13:39:28                ^
13:39:28 <stdin>:1:1: note: scanning from here
13:39:28 Handled a.swift
13:39:28 ^
```

**Comment by jrose-apple** [comment]
…Well, I don't get it either, but to not block this PR, how about you just revert the changes to test/Driver/filelists.swift for now? The test is still using symlinks so it's already marked unsupported on win32.

**Comment by jmittert** [comment]
Reverted the filelists.swift changes. Let's see if it's happier now.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test macOS platform

**Comment by jmittert** [comment]
> 
> 
> **Build failed**
> [Swift Test OS X Platform](https://ci.swift.org/job/swift-PR-osx/11224/)
> Git Sha - [ac0d9ef](https://github.com/apple/swift/commit/ac0d9ef31a8f16e34deef1e779daff838f45f5c0)

Same as the previous run, During iphone simulator tests:
`xcodebuild: error: The authorization was denied since no user interaction was possible.`

**Comment by jrose-apple** [comment]
That might have been an Xcode update. Let's try one more time. (If not, it's a test change again and we'll have to figure out why.)

@swift-ci Please test macOS

**Comment by jrose-apple** [comment]
@swift-ci Please test macOS

**Comment by compnerd** [comment]
Yay, gonna merge this, as this prevents the test harness from dying!

**Pull request merged**
